src/amici_hdf5.cpp

```diff
@@ -16,28 +17,32 @@
 #endif
 #include <unistd.h>
 
+void getModelDims(int *nx, int *nk, int *np);
+
 namespace amici {
 
 UserData *AMI_HDF5_readSimulationUserDataFromFileName(const char *fileName,
-                                                      const char *datasetPath,
-                                                      Model *model) {
+                                                      const char *datasetPath) {
 
     hid_t file_id = H5Fopen(fileName, H5F_ACC_RDONLY, H5P_DEFAULT);
 
     UserData *udata = AMI_HDF5_readSimulationUserDataFromFileObject(
-        file_id, datasetPath, model);
+        file_id, datasetPath);
 
     H5Fclose(file_id);
 
     return (udata);
 }
 
+template <class hid_t>
 UserData *AMI_HDF5_readSimulationUserDataFromFileObject(hid_t fileId,
-                                                        const char *datasetPath,
-                                                        Model *model) {
+                                                        const char *datasetPath) {
     assert(fileId > 0);
+    
+    int nx,nk,np;
+    getModelDims(&nx,&nk,&np);
```

Comment:
I added the `Model` argument so that it is clear for which model the data is to be loaded, that didn't look super nice. However, using getModelDims() makes it impossible to use this function in an application which uses two different models. I would argue that we should use getModel, getModelDims, ... nowhere other than in mexFunction and inside tests.

Reply:
currently not possible as nx,nk,np need to be available for the creation of userData, which is used for the instantiation of the model. My current approach would be to create a model-specific ReturnData class that instantiates UserData and Model. We could then give a model specific name to getModelDims and reenable the use of multiple models, but I think this should not also be adressed in this monstrosity of a pull request

Reply:
I don't see the benefit of a model-specific ReturnData class. A ReturnData class that instantiates UserData and Model feels rather unnatural. Why not the other way around? This would involve only adding a convenience function something like Model::getReturnData() { return ReturnData(this) }

My main point is that getModelDims is only defined at link time, which makes it rather hard to follow program flow and to debug stuff. Simply passing Model down there is much clearing the getting it out of nowhere.

Reply:
Currently the model only has a transient state and I would rather keep it that way. Still I agree that extending the results class might not be the best idea.  Maybe call it a simulator/kernel object that encapsulates all the other currently existing classes.

Still, currently instantiating the model requires instantiating udata which requires a call to getModelDims, so we cannot pass the model when getModelDims is called.

Reply:
Instantiating the model does not require udata.

Model::Model(const int nx, const int nxtrue,
              const int ny, const int nytrue, const int nz, const int nztrue,
              const int ne, const int nJ, const int nw, const int ndwdx,
              const int ndwdp, const int nnz, const int ubw, const int lbw,
              const AMICI_o2mode o2mode, const std::vector<realtype> p,
              const std::vector<realtype> k, const std::vector<int> plist,
              const std::vector<realtype> idlist, const std::vector<int> z2event)

Reply:
const std::vector p,
const std::vector k,
const std::vector plist
all come from udata, see respective wrapfunctions.cpp. If we do not know the size of plist at that point, we could aswell skip the initialization altogether as the size of many members depends on it.

Reply:
I see your point. Still I argue that AMICI core files should not depend on any functions that are only defined in model-specific files. 
In this function it can be replaced by reading those dimensions from the data file later on.
Maybe we should discuss (and document) the intended responsibilities and dependencies of udata, model, rdata soonish.

Reply:
agreed